internal/db/txn.go

```diff
@@ -32,26 +37,179 @@ func (t *explicitTxn) Discard(ctx context.Context) {
 
 // transactionDB is a db that can create transactions.
 type transactionDB interface {
-	NewTxn(context.Context, bool) (datastore.Txn, error)
+	NewTxn(context.Context, bool) (client.Txn, error)
 }
 
-// ensureContextTxn ensures that the returned context has a transaction
-// and an identity.
+// ensureContextTxn ensures that the returned context has a transaction.
 //
 // If a transactions exists on the context it will be made explicit,
 // otherwise a new implicit transaction will be created.
 //
-// The returned context will contain the transaction and identity
+// The returned context will contain the transaction
 // along with the copied values from the input context.
 func ensureContextTxn(ctx context.Context, db transactionDB, readOnly bool) (context.Context, datastore.Txn, error) {
 	// explicit transaction
-	txn, ok := txnctx.TryGet(ctx)
+	ctxTxn, ok := datastore.CtxTryGetTxn(ctx)
 	if ok {
-		return InitContext(ctx, &explicitTxn{txn}), &explicitTxn{txn}, nil
+		switch txn := ctxTxn.(type) {
```

Comment:
suggestion: A comment explaining this could be quite handy, I feel like this is not easily explained by the function context itself. 

`&explicitTxn{txn.Txn}` is particularly odd looking and it is not clear why this needs to be re-wrapped here.

Reply:
I removed the `explicitTxn` type.